ffi/diplomat/src/datetime_formatter.rs

```diff
@@ -0,0 +1,332 @@
+            write: &mut diplomat_runtime::DiplomatWriteable,
+        ) -> DiplomatResult<(), ICU4XError> {
+            #[allow(unused_variables)]
+            let result = self
+                .0
+                .format_to_write(write, &value.0)
+                .map_err(Into::into)
+                .into();
+            write.flush();
+            result
+        }
+
+        /// Formats a [`ICU4XIsoDateTime`] to a string.
+        #[diplomat::rust_link(icu::datetime::TimeFormatter::format_to_write, FnInStruct)]
+        #[diplomat::rust_link(icu::datetime::TimeFormatter::format, FnInStruct, hidden)]
+        #[diplomat::rust_link(icu::datetime::TimeFormatter::format_to_string, FnInStruct, hidden)]
+        pub fn format_iso_datetime(
+            &self,
+            value: &ICU4XIsoDateTime,
+            write: &mut diplomat_runtime::DiplomatWriteable,
+        ) -> DiplomatResult<(), ICU4XError> {
+            #[allow(unused_variables)]
+            let result = self
+                .0
+                .format_to_write(write, &value.0)
+                .map_err(Into::into)
+                .into();
+            write.flush();
+            result
+        }
+    }
+
+    #[diplomat::opaque]
+    /// An ICU4X TypedDateFormatter object capable of formatting a [`ICU4XGregorianDateTime`] as a string,
+    /// using the Gregorian Calendar.
+    #[diplomat::rust_link(icu::datetime::TypedDateFormatter, Struct)]
+    pub struct ICU4XGregorianDateFormatter(pub TypedDateFormatter<Gregorian>);
+
+    #[diplomat::enum_convert(length::Date, needs_wildcard)]
+    pub enum ICU4XDateLength {
+        Full,
+        Long,
+        Medium,
+        Short,
+    }
+
+    impl ICU4XGregorianDateFormatter {
+        /// Creates a new [`ICU4XGregorianDateFormatter`] from locale data.
+        #[diplomat::rust_link(icu::decimal::TypedDateFormatter::try_new_unstable, FnInStruct)]
+        pub fn try_new(
+            provider: &ICU4XDataProvider,
+            locale: &ICU4XLocale,
+            length: ICU4XDateLength,
+        ) -> DiplomatResult<Box<ICU4XGregorianDateFormatter>, ICU4XError> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+
+            let locale = locale.to_datalocale();
+
+            TypedDateFormatter::try_new_unstable(&provider, &locale, length.into())
+                .map(|df| Box::new(ICU4XGregorianDateFormatter(df)))
+                .map_err(Into::into)
+                .into()
+        }
+
+        /// Formats a [`ICU4XGregorianDateTime`] to a string.
+        #[diplomat::rust_link(icu::datetime::TypedDateFormatter::format_to_write, FnInStruct)]
+        #[diplomat::rust_link(icu::datetime::TypedDateFormatter::format, FnInStruct, hidden)]
+        #[diplomat::rust_link(
+            icu::datetime::TypedDateFormatter::format_to_string,
+            FnInStruct,
+            hidden
+        )]
+        pub fn format_datetime(
```

Comment:
Nit: There should be a `format_date` version as well

Suggestion: Should these be `format_gregorian_date[time]`, which is redundant with the class name but more consistent with the functions in the other formatter?